plugins/experimental/cookie_remap/cookiejar.cc

```diff
@@ -0,0 +1,273 @@
+      }
+    }
+    else /* If val is nullptr, then no need to add this kv pair to hashtable, skip this key */
+    {
+      continue;
+    }
+
+    /* we are going to verify the key name for our top level
+     * cookie names only so we'll use the val_check variables
+     * to know what we're processing */
+
+    if (val_check && verify_name(key) != 0)
+      continue;
+
+    if (mainElement)
+      addElement(key, addme);
+    else
+      addSubElement(key, addme);
+  }
+
+  free(arg_copy);
+  return 0;
+}
+
+void
+CookieJar::addElement(const char *key, const char *val)
+{
+  /* The insert method avoids duplicates */
+  CookieVal cval;
+  cval.m_val = val;
+  m_jar.insert(std::make_pair(key, cval));
+}
+
+void
+CookieJar::addSubElement(const char *key, const char *val)
+{
+  /* The insert method avoids duplicates */
+  m_currentVal->m_subelements.insert(std::make_pair(key, val));
+}
+
+int
+CookieJar::verify_value(char *val, int val_len)
+{
+  char *buf;
+  char *data_ptr = nullptr;
+  char data_buf[1024] = {0,};
+  int buf_len;
+  
+  if (val_len > (int) sizeof(data_buf) - 1)
+    {
+      buf_len = val_len + 1;
+      if ((data_ptr = (char *)malloc(buf_len)) == nullptr)
+	return -1;
+
+      buf = data_ptr;
+    }
+  else
+    {
+      buf = data_buf;
+      buf_len = sizeof(data_buf);
+    }
+  
+  if (get_stripped(val, val_len, buf, &buf_len, 0) != STRIP_RESULT_OK)
+    {
+      if (data_ptr)
+	free(data_ptr);
+      return -1;
+    }
+
+  /* returned buf_len includes the null terminator 
+   * so we need to verify that somehow we didn't end up
+   * with a bigger buffer than what we started with */
+
+  if (buf_len > val_len + 1)
+    {
+      if (data_ptr)
+	free(data_ptr);
+      return -1;
+    }
+
+  memcpy(val, buf, buf_len);
+  if (data_ptr)
+    free(data_ptr);
+
+  return 0;
+}
+
+int
+CookieJar::verify_name(char *name)
+{
+  char *p;
+
+  for (p = name; *p; p++)
+    {
+      /* if we get any invalid characters then return failure
+       * in order to skip this cookie completely */
+
+      if (rfc_cookie_name_table[(int) *p] == 0)
+	return -1;
+    }
+
+  return 0;
+}
+
+bool
+CookieJar::get_full(const string &cookie_name, string &val)
+{
+  if (m_jar.find(cookie_name) != m_jar.end())
+  {
+    val = m_jar[cookie_name].m_val;
+    return true;
+  }
+  return false;
+}
+
+bool
+CookieJar::get_part(const string &cookie_name, const string &part_name, string &val)
+{
+  if (m_jar.empty())
+    return false;
+
+  if (m_jar.find(cookie_name) == m_jar.end())
+  {
+    /* full cookie not found */
+    return false;
+  }
+
+  CookieVal &fe = m_jar[cookie_name];
+  /* check if we need to do lazy evaluation */
+  if (fe.parts_inited == false)
+  {
+    /* since we already validated the value for the full cookie
+     * there is no need to validate the components again so
+     * we'll be passing 0/false for the val_check argument */
+
+    m_currentVal = &fe;
+    if (parse(fe.m_val, "&", 0, false) != 0)
```

Comment:
Elsewhere this was called as `<char*, char*, bool, bool>` but now it's `<char*, char*, int, bool>`?